File: 15. 3Sum/Solution.java
public class Solution {
    public List<List<Integer>> threeSum(int[] num) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (num == null || num.length < 3)
            return res;
        
        Arrays.sort(num);
        for (int i = 0; i < num.length; i++) {
            int target = 0 - num[i];
            int l = i + 1;
            int r = num.length - 1;
            while (l < r) {
                if (num[l] + num[r] == target) {
                    List<Integer> temp = new ArrayList<Integer>();
                    temp.add(num[i]);
                    temp.add(num[l]);
                    temp.add(num[r]);
                    if (!res.contains(temp))
                        res.add(temp);
                    l++;
                    r--;
                } else if (num[l] + num[r] > target) {
                    r--;
                } else {
                    l++;
                }
            }
        }
        return res;
    }
}

File: 15. 3Sum/Solution.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        vector<vector<int>> res;
        sort(num.begin(), num.end());
        
        for (int i = 0; i < num.size(); i++) {
            int l = i+1;
            int r = num.size() - 1;
            int remain = 0 - num[i];
            
            while (l < r) {
                if (remain = num[l] + num[r]) {
                    vector<int> tmp;
                    tmp.push_back(num[i]);
                    tmp.push_back(num[l]);
                    tmp.push_back(num[r]);
                    res.push_back(tmp);
                    l++;
                    r--;
                    
                    while (l < r && num[l] == num[l-1])
                        l++;
                        
                    while (l < r && num[r] == num[r+1])
                        r--;
                    
                    if (find(res.begin(), res.end(), tmp) != res.end())
                        res.push_back(tmp);
                } else if (remain > num[l] + num[r]) {
                    l++;
                } else {
                    r--;
                }
            }
        }
        return res;
    }
};